walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,118 @@
+     field for every minute. */
+  private void setMandatoryAttendeesAvailability(ArrayList<MinuteDetails> dayMinutes,
+      Collection<Event> events, Collection<String> attendees) {
+        for (Event event : events) {
+          for (String person : attendees) {
+            if (event.attends(person)) {
+              /* Set all the minutes of the current event as unavailabe in dayMinutes. */
+              int start = event.getWhen().start();
+              int end = event.getWhen().end();
+              /* Update the next availability for the minute before the event. */
+              if (start > 0 && dayMinutes.get(start - 1).getNextAvailableMinute() < end) {
+                dayMinutes.get(start - 1).setNextAvailableMinute(end);
+              }
+              for (int i = start; i < end; i++) {
+                if (i <= TimeRange.END_OF_DAY) {
+                  dayMinutes.get(i).setMandatoryAvailability(false);
+                  /* Update the next availability for the current minute only if the setted
+                  availability does not point to an even later time. */
+                  if (dayMinutes.get(i).getNextAvailableMinute() < end) {
+                    dayMinutes.get(i).setNextAvailableMinute(end);    
+                  }
+                } else {
+                  break;
+                }
+              }
+              /* There is no point in continuing with checking if other persons are in the
+                current event, since those time slots are already set as unavailable. */ 
+              break;
+            }
+          }
+        }
   }
+
+  /* Returns a list of time ranges which fulfill all the requierments given in the request. */
+
+  private Collection<TimeRange> findAavailableTimeRanges(ArrayList<MinuteDetails> dayMinutes,
+      int duration) {
+
+      Collection<TimeRange> timeRanges = new ArrayList<>();
+      /* If the duration exceeds a day, there is no possible time range. */
+      if (duration > TimeRange.WHOLE_DAY.duration()) {
+        return timeRanges;
+      }
+      
+      int currentStart = TimeRange.START_OF_DAY;
+      /* Current start has to always be at an available spot. */ 
+      if (!dayMinutes.get(currentStart).getMandatoryAvailability()) {
+        currentStart = dayMinutes.get(currentStart).getNextAvailableMinute();
+      }
+      /* The last minute in the range must have this formula to have a range with a number
+         of minutes equal to duratio, as the number of minutes between start and end is
+         currentEnd - currentStart + 1= duration, because we want to include the  ends too. */
+      int currentEnd = currentStart + duration - 1;
+      /* This variabile keeps track if we have found other suitable ending since we started
+         analyzing the current start, it it used when current end reaches a minute that is not
+         available. */
+      int prevEnd = -1;
+      
+
+      while (currentStart < TimeRange.END_OF_DAY) {
```

Comment:
Optional: I'm wondering if this code would become slightly easiser if you only had one index to manage rather than 3 (currentStart, currentEnd, prevEnd), and compute the others on the fly.

// e.g. something like:

```
start = START_OF_DAY;
while (start < END_OF_DAY) {
  while (!start.isAvailable()) {
    start = start.getnextavailable();
  }
  end = start.getNextAvailable();
  if (end - start > length) {
    appendRange(start, end);
  }
  start = end + 1;
}
```

Reply:
The thing is for intervals where there are no unavailabilies, the getNextAvailable will send you to the exact next minute. And we want to get the maximum possible interval, not to append every time we found a good one. That's why I was keeping track on the prev, when I hit rock bottom with end, to use prev.

Reply:
Ah gotcha. I think the pseudocode above only works if instead of `getnextavailable` we implement a `getnextblock()` method.

This would return the end of the unavailability block if the current minute is unavailable or the end of the availability block if the current minute is available (thus only really storing useful information).

Let me know if you'd like to take a stab at that, or if you'd prefer to proceed with the current approach.